paasta_tools/deployd/master.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import inspect
+import logging
+import socket
+import time
+
+from six.moves.queue import Empty
+
+from paasta_tools.deployd import watchers
+from paasta_tools.deployd.common import PaastaQueue
+from paasta_tools.deployd.common import PaastaThread
+from paasta_tools.deployd.common import ServiceInstance
+from paasta_tools.deployd.leader import PaastaLeaderElection
+from paasta_tools.deployd.workers import PaastaDeployWorker
+from paasta_tools.marathon_tools import DEFAULT_SOA_DIR
+from paasta_tools.utils import get_services_for_cluster
+from paasta_tools.utils import load_system_paasta_config
+from paasta_tools.utils import ZookeeperPool
+
+log = logging.getLogger(__name__)
+logging.basicConfig(level=logging.DEBUG)
+
+
+class Inbox(PaastaThread):
+    def __init__(self, inbox_q, bounce_q):
+        PaastaThread.__init__(self)
+        PaastaThread.daemon = True
+        PaastaThread.name = "Inbox"
+        self.inbox_q = inbox_q
+        self.bounce_q = bounce_q
+        self.to_bounce = {}
+
+    def run(self):
+        while True:
+            self.process_inbox()
+        pass
+
+    def process_inbox(self):
+        try:
+            service_instance = self.inbox_q.get(block=False)
+        except Empty:
+            service_instance = None
+        if service_instance:
+            self.log.debug("Processing {} to decide if we need to add it to bounce q".format(service_instance))
+            self.process_service_instance(service_instance)
+        if self.inbox_q.empty() and self.to_bounce:
+            self.log.debug("Cleared inbox, checking if any services are ready to bounce")
+            self.process_to_bounce()
+        time.sleep(1)
+
+    def process_service_instance(self, service_instance):
+        service_instance_key = "{}.{}".format(service_instance.service, service_instance.instance)
+        if self.should_add_to_bounce(service_instance, service_instance_key):
+            self.log.debug("Adding {} to be bounced in the future".format(service_instance))
+            self.to_bounce[service_instance_key] = service_instance
+
+    def should_add_to_bounce(self, service_instance, service_instance_key):
+        if service_instance_key in self.to_bounce:
+            if service_instance.bounce_by > self.to_bounce[service_instance_key].bounce_by:
+                self.log.debug("{} already in bounce q with higher priority".format(service_instance))
+                return False
+        return True
+
+    def process_to_bounce(self):
```

Comment:
So the plan is that this gives us a way to do the "priority by time" thing and staggering of bounces.

The idea is *everything* adds to `inbox_q` and then we only move things to the real `bounce_q` if they have reached the time we want them to bounce at. We can also add logic so instances get pushed real queue if it's empty i.e. there's nothing of higher priority stopping us doing so.